decentraland/unity-renderer
# fix: Offline friends sometimes don't load

## What does this PR change?
Fix #4023 (Point 2 of this issue: `Friends who are offline are not being displayed`)

## Problem
From the second time on we opened the friends window, the offline list was always loading empty.
The problem was that, after re-opening the HUD from a second time and clean its lists, when we received the new list of friends from the backend, when the flow reached the `UpdateFriendshipStatus()` function, we had this condition:
```
if (friends.ContainsKey(userId) && friends[userId].friendshipStatus == friendshipStatus)
    return;
``` 

In this case, as the new friends received already were loaded in our internal dictionary `friends`, the flow was being always canceled and never added the data into the HUD.

The main problem here is that we were using the `UpdateFriendshipStatus()` for managing the request of friends, when this function should be used only to keep updated the status of the already loaded friends.

## Solution
In order to not depend on the `UpdateFriendshipStatus()` flow for getting the list of friends, we have made the `GetFriends()` function in the `FriendsController` as async (`GetFriends()` -> `GetFriendsAsync()`), then the consumer (`FriendsHUDController`) can call it and wait for the response to directly fill the list in the UI.

## How to test the changes?
1. Go to: https://play.decentraland.zone/?renderer-branch=fix/offline-friends-are-not-being-displayed
2. Open the friends list by first time.
3. Check the friends list is loaded correctly.
4. Close the friends list.
5. Open it again.
6. Check the friends list is loaded correctly.
7. Repeat this process several times and check that the lists are being loaded correctly.
8. Check the friends feature is working as expected.

## Important
This has been the quicker solution in order to add the fix in the next release and let the users load their offline friends list correctly, but as a second step we should synchronize with dService team in order to create a new RPC function for the current `GetFriends` approach. With these changes, it should be something very easy to adapt.

## Our Code Review Standards
https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md
